[kibana] support api_key

sachin-frayne opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Summary

Adds support for api_key authentication in the kibana module

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally
- create an api key and grab the beats format[1]
- configure kibana module to use that api_key:
```
metricbeat.modules:
  - module: kibana
    xpack.enabled: true
    hosts: ["http://localhost:9200"]
    api_key: <key in beats format>
    period: 10s
```
- kibana metricsets are successfully collecting documents

[1]
![Screenshot 2023-08-09 at 17 23 25](https://github.com/elastic/beats/assets/5239883/1008feb8-afa9-4e90-a47c-8d0c1379a8ed)

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

- Step one to adding api_key to Kibana module: https://github.com/elastic/enhancements/issues/22286

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

klacabane commented on metricbeat/module/kibana/cluster_actions/cluster_actions.go:
```diff
@@ -88,7 +88,7 @@ func (m *MetricSet) Fetch(r mb.ReporterV2) (err error) {
 
 func (m *MetricSet) validate() (err error, versionSupported bool) {
-	kibanaVersion, err := kibana.GetVersion(m.actionsHTTP, kibana.ClusterActionsPath)
+	kibanaVersion, err := kibana.GetVersion(m.actionsHTTP, kibana.ClusterActionsPath, m.ApiKey)
```
Can we append the api key to the http clients when we init the metricsets so we don't have to worry about passing the key around ? Something similar to what the [elasticsearch module is doing](https://github.com/elatic/beats/blob/main/metricbeat/module/elasticsearch/metricset.go#L114) although I'm not sure if it's feasible here

sachin-frayne replied:
I am also not sure if it is feasible here. I tried a couple of ways, but in the end passing it through like this ended up being the easiest way for me to get it working.

VihasMakwana commented:
@sachin-frayne can you take care of CI failures. LGTM otherwise.

VihasMakwana commented:
I've rebased main.

sachin-frayne commented:
> I've rebased main.

Thank you

sachin-frayne commented:
Hello @VihasMakwana.

I have fixed the CI, I have also changed the setting in: [.github/workflows/golangci-lint.yml](https://github.com/elastic/beats/pull/40360/files#diff-d9a956120ac84289d2650137f64bd8f0893331de05e44cc41899dd984c9e0d48) removing: 

```skip-go-installation: true```

I don't think this was causing any failures, but it debugging I found that the setting is removed/being removed: https://github.com/golangci/golangci-lint-action

Can you please review one more time?

sachin-frayne commented:
Hello @VihasMakwana 

When you get a chance can you please review this?

pierrehilbert commented:
@alexsapran / @pazone could you please review here?

alexsapran commented:
The task is to review this line.
https://github.com/elastic/beats/pull/40360/files#diff-d9a956120ac84289d2650137f64bd8f0893331de05e44cc41899dd984c9e0d48

right @pierrehilbert?

alexsapran commented on .github/workflows/golangci-lint.yml:
```diff
@@ -53,8 +53,5 @@ jobs:
           args: --timeout=30m --whole-files
 
-          # Optional: if set to true then the action will use pre-installed Go.
-          skip-go-installation: true
```
This was deprecated in 

```
v4.0.0+ requires an explicit actions/setup-go installation step before using this action: uses: actions/setup-go@v5. The skip-go-installation option has been removed.
```

But we are running version `v1.55.2`, so I am not sure removing this achieves what you want.
Why we wan to remove this line?

sachin-frayne replied:
I will revert this change now.

pierrehilbert commented:
Yes this is where your review is needed, the rest of the PR has been reviewed by @belimawr

sachin-frayne commented:
@pierrehilbert and @belimawr

Since I have reverted the change with the `skip-go-installation` can this move onto the next stage, I believe I am waiting for a reviewer approval.

Thank you.

belimawr commented on metricbeat/module/kibana/kibana.go:
```diff
@@ -159,5 +159,6 @@ func fetchPath(http *helper.HTTP, currentPath, newPath string) ([]byte, error) {
 
 	// Http helper includes the HostData with username and password
 	http.SetURI(u.String())
+	http.SetHeader("Authorization", "ApiKey "+apiKey)
```
[Question]
Should this header be always set or only when `apiKey` is present (not an empty string)?

belimawr commented on metricbeat/module/kibana/status/status.go:
```diff
@@ -72,6 +72,8 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 // It returns the event which is then forward to the output. In case of an error, a
 // descriptive error must be returned.
 func (m *MetricSet) Fetch(r mb.ReporterV2) error {
+	apiKey := m.ApiKey
+	m.http.SetHeader("Authorization", "ApiKey "+apiKey)
```
[Question]
Same question here: should this always be set or only when `apiKey` is not empty?

Also as a suggestion:
```suggestion
	m.http.SetHeader("Authorization", "ApiKey "+m.ApiKey)
```

sachin-frayne commented:
Hey @belimawr, :)

Thank you for your review and the suggestions, you are correct the header should only be set if the apikey is not an empty string, I have implemented an if statement around the function.

Please review and let me know if you are happy with the changes?

sachin-frayne commented:
Hello @VihasMakwana:

I am not sure, but I believe I am still waiting for your approval.

pierrehilbert commented:
@sachin-frayne as @VihasMakwana and @belimawr are in the same team, one of them you should be enough. 
Looks like you need someone from @elastic/stack-monitoring

consulthys commented:
@sachin-frayne is @elastic/stack-monitoring :-)

pierrehilbert commented:
> @sachin-frayne is @elastic/stack-monitoring :-)

I still think you need someone from your team to approve the change according to the codeowners.
